pkg/ingester/wal.go

```diff
@@ -384,35 +387,96 @@ func recoverFromWAL(ingester *Ingester) (err error) {
 		seriesCache[i] = make(map[string]map[uint64]*memorySeries)
 	}
 
-	if idx >= 0 {
-		// Checkpoint exists.
-		level.Info(util.Logger).Log("msg", "recovering from checkpoint", "checkpoint", lastCheckpointDir)
-		start := time.Now()
-		if err := processCheckpoint(lastCheckpointDir, userStates, nWorkers, stateCache, seriesCache); err != nil {
-			return err
-		}
-		elapsed := time.Since(start)
-		level.Info(util.Logger).Log("msg", "recovered from checkpoint", "time", elapsed.String())
-	} else {
-		level.Info(util.Logger).Log("msg", "no checkpoint found")
+	level.Info(util.Logger).Log("msg", "recovering from checkpoint")
+	start := time.Now()
+	userStates, idx, err := processCheckpointWithRepair(walDir, ingester, nWorkers, stateCache, seriesCache)
+	if err != nil {
+		return err
 	}
+	elapsed := time.Since(start)
+	level.Info(util.Logger).Log("msg", "recovered from checkpoint", "time", elapsed.String())
 
 	if segExists, err := segmentsExist(walDir); err == nil && !segExists {
 		level.Info(util.Logger).Log("msg", "no segments found, skipping recover from segments")
+		ingester.userStatesMtx.Lock()
+		ingester.userStates = userStates
+		ingester.userStatesMtx.Unlock()
 		return nil
+	} else if err != nil {
+		return err
 	}
 
 	level.Info(util.Logger).Log("msg", "recovering from WAL", "dir", walDir, "start_segment", idx)
-	start := time.Now()
-	if err := processWAL(walDir, idx, userStates, nWorkers, stateCache, seriesCache); err != nil {
+	start = time.Now()
+	if err := processWALWithRepair(walDir, idx, userStates, nWorkers, stateCache, seriesCache, ingester.metrics); err != nil {
 		return err
 	}
-	elapsed := time.Since(start)
+	elapsed = time.Since(start)
 	level.Info(util.Logger).Log("msg", "recovered from WAL", "time", elapsed.String())
 
+	ingester.userStatesMtx.Lock()
+	ingester.userStates = userStates
+	ingester.userStatesMtx.Unlock()
 	return nil
 }
 
+func processCheckpointWithRepair(walDir string, ingester *Ingester, nWorkers int,
+	stateCache []map[string]*userState, seriesCache []map[string]map[uint64]*memorySeries) (*userStates, int, error) {
+
+	// Use a local userStates, so we don't need to worry about locking.
+	userStates := newUserStates(ingester.limiter, ingester.cfg, ingester.metrics)
```

Comment:
Isn't this going to affect real ingester metrics? Is that a problem?

Reply:
I will check about that, good catch

Reply:
Cross checked with how transfer handles them https://github.com/cortexproject/cortex/blob/5d7b05c36bbec3e2f478e924c3e1e1a853886abf/pkg/ingester/transfer.go#L65-L151

Looks like it's the same here too (including the memory chunks metrics). So I guess all good here.